EducationalAssignments/UndoPartOne.md

```diff
@@ -0,0 +1,303 @@
+# Implement a Defensive Security System
+
+This assignment will help you understand security mechanisms. You will be guided
+through the steps of creating a reference monitor using the security layer
+functionality in Repy V2. A reference monitor is an access control concept that
+refers to an abstract machine that mediates all access to objects by subjects.
+This can be used to allow, deny, or change the behavior of any set of calls.
+While not a perfect way of validating your reference monitor, it is useful to
+create test cases to see whether your security layer will work as expected (the
+test cases may be turned in as part of the next assignment). 
+
+This assignment is intended to reinforce concepts about access control and
+reference monitors in a hands-on manner. 
+
+
+## Overview
+
+In this assignment, you will implement a defense monitor to oversee file
+operations in Repy. The monitor will enhance the default Repy behavior by adding
+an undo functionality for writes. This defense monitor will ensure that software
+follows certain rules and guidelines, preventing potential unauthorized actions.
+
+The focus is on monitoring file write operations and allowing the last write to
+be reverted. This undo capability mimics common document editors that let you
+reverse your most recent edit.
+
+You should write test applications to ensure your reference monitor behaves
+properly in different cases and to test attacks against your monitor.    
+
+
+### Specifications:
+
+1. Your defense monitor should have all the standard file operation methods plus
+   an additional method called `undo`.
+2. The writeat operation will not immediately reflect its changes on the file.
+   Instead, the changes will only become permanent after:
+    - A subsequent valid `writeat` operation is executed.
+    - The file is closed.
+3. The `undo` operation will reverse the last `writeat` operation's changes. If
+   multiple `writeat` operations are performed without any of them being
+   finalized (i.e., without an intervening `writeat`, `close`, or `undo`), only
+   the last `writeat` is affected by the `undo`. If there haven't been any
```

Comment:
There cannot be multiple pending writeat operations, correct?  Since the rule above says that only valid writeats can be pending and another valid writeat will cause the first to be written

Reply:
I'm sorry, I think we discussed this before. I probably had copied a previous draft in the PR. I updated the instructions.

Reply:
okay, please go through and be sure all other edits were also applied.